internal/flavors/publisher_test.go

```diff
@@ -208,3 +210,140 @@ func sleepOnCycleEnd(t *testing.T, size int, eventCount int, interval time.Durat
+			threshold:                           3,
+			incomeEventBatchesOf:                []int{5},
+			expectedPublishedBatchesIDs:         [][]int{{1, 2, 3, 4, 5}},
+			expectedBufferCapacityOnEachPublish: []int{8},
+			expectedBufferCapacityEnd:           8, // 4 * 2
+		},
+		"single batch over threshold break capacity limit": {
+			interval:                            2 * time.Second,
+			threshold:                           3,
+			incomeEventBatchesOf:                []int{15},
+			expectedPublishedBatchesIDs:         [][]int{{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15}},
+			expectedBufferCapacityOnEachPublish: []int{15},
+			expectedBufferCapacityEnd:           4,
+		},
+		"multiple batches break capacity limit": {
+			interval:                            2 * time.Second,
+			threshold:                           5,
+			incomeEventBatchesOf:                []int{2, 2, 20},
+			expectedPublishedBatchesIDs:         [][]int{{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24}},
+			expectedBufferCapacityOnEachPublish: []int{27},
+			expectedBufferCapacityEnd:           7,
+		},
+	}
+
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			log := testhelper.NewLogger(t)
+			defer goleak.VerifyNone(t, goleak.IgnoreCurrent())
+
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			var publisher *Publisher
+
+			client := newMockClient(t)
+			calls := make([]*mock.Call, len(tc.expectedPublishedBatchesIDs))
+			for i, publishedSliceIDs := range tc.expectedPublishedBatchesIDs {
+				calls[i] = client.EXPECT().PublishAll(events(publishedSliceIDs)).Run(
+					func(_ []beat.Event) {
+						assert.Equalf(t, tc.expectedBufferCapacityOnEachPublish[i], cap(publisher.eventsBuffer), "buffer capacity miss-match on publish number %d", i)
+					},
+				).Call
+			}
+			mock.InOrder(calls...)
+
+			publisher = NewPublisher(log, tc.interval, tc.threshold, client)
+			eventsChannel := make(chan []beat.Event, 10)
+
+			// send events
+			go func() {
+				events := generateEventBatches(tc.incomeEventBatchesOf...)
+				for _, batch := range events {
+					eventsChannel <- batch
+				}
+				close(eventsChannel)
+			}()
+
+			publisher.HandleEvents(ctx, eventsChannel)
+			require.Equal(t, tc.expectedBufferCapacityEnd, cap(publisher.eventsBuffer))
+		})
+	}
+}
+
+func generateEventBatches(batchesOf ...int) [][]beat.Event {
+	batches := make([][]beat.Event, 0, len(batchesOf))
+
+	id := 0
+
+	for _, singleBatchCount := range batchesOf {
+		singleBatch := make([]beat.Event, 0, singleBatchCount)
+		for range singleBatchCount {
+			id++
+			singleBatch = append(singleBatch, event(id))
+		}
+		batches = append(batches, singleBatch)
+	}
+
+	return batches
+}
+
+func event(id int) beat.Event {
+	return beat.Event{
+		Fields: mapstr.M{"id": id},
+	}
+}
+
+func events(ids []int) []beat.Event {
+	s := make([]beat.Event, len(ids))
+	for i := range len(ids) {
+		s[i] = event(ids[i])
+	}
+	return s
+}
```

Comment:
```suggestion
func generateEventBatches(batchesOf ...int) [][]beat.Event {
	batches := make([][]beat.Event, 0, len(batchesOf))

	id := 1
	for _, singleBatchCount := range batchesOf {
		batches = append(batches, eventsFromIds(lo.RangeFrom(id, singleBatchCount)))
		id += singleBatchCount
	}

	return batches
}

func eventsFromIds(ids []int) []beat.Event {
	s := make([]beat.Event, len(ids))
	for i := range len(ids) {
		s[i] = beat.Event{
			Fields: mapstr.M{"id": ids[i]},
		}
	}
	return s
}
```

Reply:
resolving because I refactored and made them inline